### DISCUSSION TITLE
Is there a way to call an async function without using await?

### DISCUSSION THREAD

**Discussion by dsblank** [pyodide/pyodide#1703] [Q&A] [answered]
I have a basic question, but I am having trouble with pyodide's loop and asyncio functionality.

Consider this simple code:

```python
async def function():
  await asyncio.sleep(1)
  return 42

x = await function()
```

Is there a way I can call `function()` getting 42, without using the word `await`? I've tried everything I can think of but everything requires an `await` at the top of the calling stack. I don't mind if it blocks other coroutines.

**Comment by RainMa22** [comment]
have you tried
`import asyncio`
`x = asyncio.run(function())`
?

**Reply by rth** [reply]
No this wouldn't work, I think, because `asyncio.run` attempts to create a new event loop instead of using the browser one. The error is,
```py
>>> x = asyncio.run(function())
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/lib/python3.9/asyncio/runners.py", line 33, in run
    raise RuntimeError(
RuntimeError: asyncio.run() cannot be called from a running event loop
```

**Comment by rth** [comment]
You could use [`asyncio.loop.run_until_complete`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_until_complete),
```py
# get the browser event loop
loop = asyncio.get_event_loop()
task = loop.run_until_complete(function())
task.result()  # returns 42
```

**Reply by hoodmane** [reply]
I am pretty sure this won't work either and that there is no way to do this. We can't block for the event loop because that blocks the current running thread which prevents the browser event loop from running. This is basically the same as #1503. On a normal Python the event loop would run on a different thread so one thread can block while the other keeps going.

**Reply by rth** [reply]
You are right. That relied on the latency of entering those lines one by one in the REPL. With one line,
```py
>>> asyncio.get_event_loop().run_until_complete(function()).result()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/lib/python3.9/asyncio/futures.py", line 198, in result
    raise exceptions.InvalidStateError('Result is not ready.')
asyncio.exceptions.InvalidStateError: Result is not ready.
```

**Comment by hoodmane** [comment, marked as answer]
Unfortunately this is not currently possible. 

It will not be possible until we get some sort of WASM_PTHREADS support. WASM_PTHREADS currently interacts badly with a bunch of other emscripten features we need, so it is unclear on what time scale this will happen if ever, but it is safe to assume that it is on the order of years.
